tests/ui/same_item_push.rs

```diff
@@ -111,4 +113,15 @@ fn main() {
     for _ in 0..10 {
         vec15.push(Box::new(S {}));
     }
+
```

Comment:
We should add simplified versions of the problematic cases of #5985 
* My other comment about shadowing should cover the first case involving a function call
* We could add a test in which the variable is initialized with a match expression.